src/Nethermind/Nethermind.Taiko/Rpc/TaikoGasPriceOracle.cs

```diff
@@ -0,0 +1,130 @@
+// SPDX-FileCopyrightText: 2025 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using Nethermind.Blockchain.Find;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Specs;
+using Nethermind.Evm;
+using Nethermind.Int256;
+using Nethermind.JsonRpc.Modules.Eth.FeeHistory;
+using Nethermind.JsonRpc.Modules.Eth.GasPrice;
+using Nethermind.Logging;
+using System.Linq;
+
+namespace Nethermind.Taiko.Rpc;
+
+public class TaikoGasPriceOracle : IGasPriceOracle
+{
+    private readonly IBlockFinder _blockFinder;
+    private readonly ISpecProvider _specProvider;
+    private readonly IFeeHistoryOracle _feeHistoryOracle;
+    private readonly ILogger _logger;
+    internal PriceCache _gasPriceEstimation;
+    private readonly UInt256 _minGasPrice;
+
+    // Constants required for computing the gas price
+    private static readonly UInt256 L2GasPerL2Batch = 1_000_000;
+    private const ulong ProvingCostPerL2Batch = 800_000_000_000_000;
+    private const ulong BatchPostingGasWithoutCallData = 180_000;
+    private const ulong BatchPostingGasWithCallData = 260_000;
+    private const ulong ProofPostingGas = 750_000;
+    private const int FeeHistoryBlockCount = 200; // Number of blocks to consider for average base fee
+    private const int BlobSize = 128 * 1024;
+
+    public TaikoGasPriceOracle(
+        IBlockFinder blockFinder,
+        ISpecProvider specProvider,
+        IFeeHistoryOracle feeHistoryOracle,
+        ILogManager logManager,
+        UInt256 minGasPrice)
+    {
+        _blockFinder = blockFinder;
+        _specProvider = specProvider;
+        _feeHistoryOracle = feeHistoryOracle;
+        _logger = logManager.GetClassLogger();
+        _minGasPrice = minGasPrice;
+    }
+
+    private UInt256 FallbackGasPrice() => _gasPriceEstimation.LastPrice ?? _minGasPrice;
+
+    public UInt256 GetGasPriceEstimate()
+    {
+        Block? headBlock = _blockFinder.Head;
+        if (headBlock is null)
+        {
+            if (_logger.IsTrace) _logger.Trace("[TaikoGasPriceOracle] No head block available, using fallback gas price");
+            return FallbackGasPrice();
+        }
+
+        Hash256 headBlockHash = headBlock.Hash!;
+        if (_gasPriceEstimation.TryGetPrice(headBlockHash, out UInt256? price))
+        {
+            if (_logger.IsTrace) _logger.Trace($"[TaikoGasPriceOracle] Using cached gas price estimate: {price}");
+            return price!.Value;
+        }
+
+        UInt256 l1BaseFee = headBlock.BaseFeePerGas;
```

Comment:
how come `L1BaseFee` equals L2 head block baseFee?

Reply:
Ah, you made a mistake with BlockFinder, block finder is the L2 block tree not the L1 block tree. to get the L1 costs you need to call the L1 RPC. you will not be able to get it locally without an external RPC call